maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,253 @@
+
+        Returns:
+            dict: Metrics information.
+        """
+
+        return DocableDict(
+            metrics_desc,
+            energy_consumption=self._energy_consumption,
+            success_requirements=self._success_requirements,
+            failed_requirements=self._failed_requirements,
+            total_latency=self._total_latency
+        )
+
+    def _register_events(self):
+        # Register our own events and their callback handlers.
+        self._event_buffer.register_event_handler(Events.REQUIREMENTS, self._on_vm_required)
+        self._event_buffer.register_event_handler(Events.FINISHED, self._on_vm_finished)
+
+        # Generate decision event.
+        self._event_buffer.register_event_handler(MaroEvents.TAKE_ACTION, self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs memory utilization.
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for _, vm in self._vm.items():
+            vm.util_cpu = vm.get_util(cur_tick=self._tick)
+
+    def _update_pm_util(self):
+        """Update memory utilization occupied by total VMs on each PM."""
+        for pm in self._machines:
+            total_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._vm[vm_id]
+                cur_util_cores = vm.util_cpu * vm.req_cpu / 100
+                total_cpu += cur_util_cores
+            pm.util_cpu = total_cpu / pm.cap_cpu * 100
+            pm.update_util_series(self._tick)
+
+        # TODO: Energy comsumption update.
+
+    def _on_vm_required(self, evt: CascadeEvent):
+        """Callback when there is a VM requirement generated."""
+        # Get VM data from payload.
+        payload: VmRequirementPayload = evt.payload
+        vm_req: VirtualMachine = payload.vm_req
+        buffer_time: int = payload.buffer_time
+
+        # Check all valid PMs.
+        # NOTE: Should we implement this logic inside the action scope?
+        valid_pm_list = [
+            pm.id
+            for pm in self._machines
+            if (pm.cap_cpu - (pm.cap_cpu * pm.util_cpu / 100)) >= vm_req.req_cpu
+        ]
+
+        if len(valid_pm_list) > 0:
+            # Generate pending decision.
+            decision_payload = DecisionEvent(
+                valid_pm=valid_pm_list,
+                vm_info=vm_req,
+                buffer_time=buffer_time
+            )
+            pending_decision_evt = self._event_buffer.gen_decision_event(evt.tick, payload=decision_payload)
+            evt.add_immediate_event(pending_decision_evt)
+            self._pending_action_vm_id = vm_req.vm_id
+        else:
+            # Postpone the buffer duration ticks by config setting.
+            if buffer_time > 0:
+                postpone_payload = payload
+                postpone_payload.buffer_time -= self._delay_duration
```

Comment:
compare the ``payload.remaining_buffer_time - self._delay_duration`` with ``0`` immediately. if it <= 0, no need to generate the postponed event and can update the ``self._failed_requirements`` right now